src/omv/imlib/imlib.h

```diff
@@ -66,6 +66,25 @@
 #define IM_DIV(a, b)             ({ __typeof__ (a) _a = (a); __typeof__ (b) _b = (b); _b ? (_a / _b) : 0; })
 #define IM_MOD(a, b)             ({ __typeof__ (a) _a = (a); __typeof__ (b) _b = (b); _b ? (_a % _b) : 0; })
 
+// IM_UCLAMPZ  - For when the value may be negative but not larger than the max.
+// IM_UCLAMPM  - For when the value may be larger than the max but not negative.
+// IM_UCLAMP   - For when the value may be negative or larger than the max.
+// IM_SCLAMP   - For when the value may be negative or larger than the max.
+// IM_ABSDIFF8 - Absolute difference between two 8-bit unsigned values.
+#ifdef ARM_MATH_DSP
+#define IM_UCLAMPZ(x, y)         __USAT((x), (y))
+#define IM_UCLAMPM(x, y)         __USAT((x), (y))
+#define IM_UCLAMP(x, y)          __USAT((x), (y))
+#define IM_SCLAMP(x, y)          __SSAT((x), (y))
+#define IM_ABSDIFF8(x, y)        __USAD8((x), (y))
+#else
+#define IM_UCLAMPZ(x, y)         IM_MAX((x), 0)
+#define IM_UCLAMPM(x, y)         IM_MIN((x), ((1 << (y)) - 1))
+#define IM_UCLAMP(x, y)          IM_MAX(IM_MIN((x), ((1 << (y)) - 1)), 0)
+#define IM_SCLAMP(x, y)          IM_MAX(IM_MIN((x), ((1 << ((y) - 1)) - 1)), -(1 << ((y) - 1)))
+#define IM_ABSDIFF8(x, y)        abs((x) - (y))
+#endif
```

Comment:
> When IM_UCLAMPZ is used for DSP/Non-DSP you don't want to clamp to zero you want to saturate to 8, __USAT(x, 8) is Not equivalent to IM_MAX(x, 0).

Yeah, as mentioned previously, the use of SIMD instructions is situation-specific. They don't map to generic macros nicely. However, in this case, because you still have to pass the expected maximum number of bits, it is safe to use.

> I realized those should actually be called IM_xSATy, especially that they're called with bit count, similarly absdiff should be IM_USAD8.

I don't think the SAT name applies in the cases of Z/M because I'm not necessarily saturating there. I can change the name, though, from CLAMP to SAT. Regarding the __USAD8. This is a special situation-specific use case of the instruction. __USAD8 sums the absolute difference between two 32-bit registers broken up into 8-bit chunks. So, in the special case of two registers having only an 8-bit unsigned value in them, it is equivalent to abs(x - y). Otherwise it's not the same. It should not be named IM_USAD8 as it's not doing that instruction.

> I don't think we need other than IM_SSAT and IM_USAT. For example IM_UCLAMPZ should be IM_SCLAMP(x, 8) it will be called with a signed value ex IM_SCLAMP((int32) a - b, 8)

In the case where the DSP instructions are not used you'd be explicitly adding uneeded code when you fallback to the MIN and MAX stuff. Note, __USAT and __SSAT both work on signed values for their input. __USAT clamps to zero though while __SSAT doesn't.

> We should also add a non-DSP general clamp macro that accepts (value, min, max) to replace all of the IM_MIN(MAX(....)) in the library.

I agree for when you are dealing with non-power of 2 values. Otherwise, the clamp is better to use. Not sure how many times we clamp when not using a power of 2.

> The comments should be the ranges (0 <= x <= ) see example below

Sure.

...

Executing changes.

Reply:
>> They don't map to generic macros nicely. However, in this case, because you still have to pass the expected maximum number of bits, it is safe to use.

The expected maximum number of bits is not used in the non-DSP version, if it's called from somewhere else it will return the wrong value (IM_MAX(x, 0) basically is x for unsigned values, but x can be > max bits. The DSP version should look like:
```C
#define IM_USAT(x, y)         __USAT((x), (y))
#define IM_SSAT(x, y)         __SSAT((x), (y))

```

The non-DSP version should look like:
```C
#define IM_USAT(x, y)        IM_MIN(x,  ((1 << (y)) - 1))
#define IM_SSATx, y)         IM_CLAMP(x, -((1 << (y)) - 1), ((1 << (y)) - 1)) // clamp is (value, min, max) macro.
```
>> I don't think the SAT name applies in the cases of Z/M because

We don't need Z/M/ only SSAT and USAT.

Reply:
Also all of those intrinsic functions have fallbacks for architectures that don't support them, see `hal/cmsis/include/cmsis_gcc.h` we don't even need to define them like that.